### ISSUE TITLE
Static files generated before build are not registered by the module static file provider on the first build

### ISSUE THREAD

**Issue by barthamark** [OrchardCMS/OrchardCore#3865]
I have a BeforeTargets="BeforeBuild" script defined in my module's csproj file that executes a Gulp task which generates a bunch of js files taking more than 4 seconds to finish. These files are not accessible by the web application only after rebuilding the solution. However, if the script finishes quickly (couple milliseconds) then it will be fine after the first build.

```
<Target Name="NpmInstall" BeforeTargets="BeforeBuild">
  <Exec Command="npm install" />
  <Exec Command="gulp" />
</Target>
```

**Comment by jtkech** [comment]
Try to do it before our target where we use `EmbeddedResource` and remove duplicates.

    <Target Name="NpmInstall" BeforeTargets="EmbeddModuleAssets">
      <Exec Command="npm install" />
      <Exec Command="gulp" />
    </Target>

Hmm, maybe the `EmbeddedResource` need to be re-evaluated, if so (and only in this case), and assuming that the gulp outputs are all under the module `wwwroot`, try this

    <Target Name="NpmInstall" BeforeTargets="UpdateEmbeddedResource">
      <Exec Command="npm install" />
      <Exec Command="gulp" />
    </Target>

    <Target Name="UpdateEmbeddedResource" BeforeTargets="EmbeddModuleAssets">
      <ItemGroup>
        <EmbeddedResource Include="wwwroot\**" Exclude="$(ModuleDefaultAssetExcludes);$(DefaultItemExcludes)" />
      </ItemGroup>
    </Target>

**Comment by sebastienros** [comment]
fixed?

**Comment by Piedone** [comment]
I haven't really seen this cause an issue lately. @barthamark?

**Comment by barthamark** [comment]
So what we've done (and created [library](https://github.com/Lombiq/npm-targets) for this and also see [this file and MSBuild target](https://github.com/Lombiq/NPM-Targets/blob/dev/Lombiq.Npm.Targets.targets#L29)) is to basically run `npm install` and `gulp` before `EmbeddModuleAssets` exactly how @jtkech suggested (except for the Update target).

Since it's working and stable we can say that it's "fixed", however, I think it would be a great idea to add this to the documentation. Also maybe fix the typo in `EmbeddModuleAsset` target name but note that it will cause breaking changes.

**Comment by BenedekFarkas** [comment]
To those finding this issue later, the mentioned build target was renamed to `OrchardCoreEmbedModuleAssets` in 9af60180f99efe5e7c2ddc42fae90647f957df7f and was first released version 1.4.0.
